src/gui/core/window_builder.hpp

```diff
@@ -116,6 +115,31 @@ builder_widget_ptr build_widget(const config& cfg)
 	return std::make_shared<T>(cfg);
 }
 
+/**
+ * Helper function to implement @ref build_single_widget_instance. This keeps the main
+ * logic in the implementation despite said function being a template and therefor
+ * needing to be fully implemented in the declaration.
+ */
+widget* build_single_widget_instance_helper(const std::string& type, const config& cfg);
+
+/**
+ * Builds a single widget instance of the given type with the specified attributes.
+ *
+ * This should be used in place of creating a widget object directly, as it
+ * allows the widget-specific builder code to be executed.
+ *
+ * @tparam                        The final widget type. The widget pointer will be
+ *                                cast to this.
+ *
+ * @param type                    String ID of the widget type.
+ * @param cfg                     Data config to pass to the widget's builder.
+ */
+template<typename T>
+T* build_single_widget_instance(const std::string& type, const config& cfg = config())
```

Comment:
Hmm... `build_widget_instance`? Or even just `build_widget`? I was also entertaining the idea of making it an `operator new` with additional arguments (ie, like placement new), not sure if that can work though.